dev/src/bulk-writer.ts

```diff
@@ -618,18 +697,38 @@ export class BulkWriter {
    *   console.log('Executed all writes');
    * });
    */
-  async flush(): Promise<void> {
+  flush(): Promise<void> {
     this.verifyNotClosed();
-    const trackedBatches = this.batchQueue;
-    const writePromises = trackedBatches.map(batch => batch.awaitBulkCommit());
-    this.sendReadyBatches();
-    await Promise.all(writePromises);
+
+    // Copy the pending ops at the time flush() was called.
+    return this._flush(Array.from(this._pendingOps));
+  }
+
+  private async _flush(pendingOps: Array<Promise<void>>): Promise<void> {
+    let batchQueue = this._batchQueue;
+    batchQueue.forEach(batch => batch.markReadyToSend());
+
+    // Send all scheduled operations on the BatchQueue first.
+    this.sendReadyBatches(batchQueue);
+    await Promise.all(this._pendingBatches);
+
+    // Afterwards, send all accumulated retry operations. Wait until the
+    // retryBatchQueue is cleared. This way, operations scheduled after
+    // flush() will not be sent until the retries are completed.
+    batchQueue = this._retryBatchQueue;
+    if (batchQueue.length > 0) {
+      this.sendReadyBatches(batchQueue);
+    }
```

Comment:
I wonder if this logic can be combined. The logic for the two different lists is essentially the same.